### PR TITLE
:bug: [JENKINS-75171] Fix ambiguous 12AM/12PM time format displayed on Build History Widget

### PR DESCRIPTION
**Pull request by SheikhSuhail19**
This PR resolves [JENKINS-75171](https://issues.jenkins.io/browse/JENKINS-75171) by adding new `getDisambiguatedTime` method to correctly display "12:00 midnight" for 12AM and "12:00 noon" for 12PM to avoid confusion.

### Before
- 12:00PM → 12:00 PM (ambiguous)
- 12:00AM → 12:00 AM (ambiguous)

### After
- 12:00PM → 12:00 noon ✅
- 12:00AM → 12:00 midnight ✅

See [JENKINS-75171](https://issues.jenkins.io/browse/JENKINS-75171).

### Testing done

- Added unit tests in `FunctionsTest.java` to cover all edge cases related to disambiguating time strings, including:
  - Normal AM/PM times (e.g., `9:15 AM`)
  - Noon (`12:00 noon`)
  - Midnight (`12:00 midnight`)
  - 24-hour formatted times (e.g., `13:30`, `23:59`)
  - One second after midnight (`12:00:01 AM`)
  - One second after noon (`12:00:01 PM`)
  - Time in different time zones (e.g., converting `12:00 UTC` to a different local time zone)
  - Handling of Daylight Saving Time (e.g., transitioning from 2:00 AM to 3:00 AM)
- Verified that all new and existing unit tests pass locally using `mvn test`
- Deployed the changes on a local Jenkins instance and verified that time values are correctly disambiguated in the UI
- Confirmed correct rendering and labeling of edge cases like `12:00 midnight` and `12:00 noon`

## Before
12am

## After
1200midnight
1200noon

image

### Proposed changelog entries

- Bug Fix: Disambiguate 12:00 AM and 12:00 PM in human-readable time format to display as “12:00 midnight” and “12:00 noon on the Build History Widget”.
- 

### Proposed changelog category

/label bug, web-ui

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, JENKINS-75171, is well-described and linked in the PR description.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change and are in the imperative mood.
- [x] There is automated testing and visual verification by deploying and validating the UI.
- [x] No new public classes, fields, or methods introduced that require `@Restricted` or `@since TODO` annotations.
- [x] No new deprecations introduced.
- [x] No new JavaScript changes or use of `eval`.
- [x] No dependency updates in this change.
- [x] No new APIs or extension points introduced.

### Desired reviewers

@jtnord
@jglick 
@alanharder 

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/Functions.java [hunk 1]
```diff
@@ -736,6 +736,44 @@ public static String getUserTimeZone() {
...
+        int hour = calendar.get(Calendar.HOUR_OF_DAY);
+        int minute = calendar.get(Calendar.MINUTE);
+
+        if (minute == 0) {
+            if (hour == 0) {
+                return Messages.DisambiguatedTime_Midnight();
+            } else if (hour == 12) {
+                return Messages.DisambiguatedTime_Noon();
+            }
+        }
```
Comment (jtnord) [inline core/src/main/java/hudson/Functions.java:769]: you only wan tto do this with TimeZones that do not use the 24 hour clock.  anything using the 24 hour clock (most of Europe for example) should say 12:00 as that is always noon (00:00 would be midnight and there is no confusion between the two).

**Review by jtnord — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/Functions.java (hunk 1 above)
Comment (SheikhSuhail19) [reply core/src/main/java/hudson/Functions.java]: Thank you for your insightful feedback. I have updated the implementation to conditionally replace 'AM'/'PM' with 'Midnight'/'Noon' only for locales utilizing the 12-hour clock format. For locales that employ the 24-hour clock, the time will continue to display as '00:00' for midnight and '12:00' for noon, ensuring clarity and consistency across different regional settings.

File: core/src/main/resources/hudson/widgets/HistoryWidget/entry.jelly [hunk 2]
```diff
@@ -52,7 +52,7 @@ THE SOFTWARE.
             <j:set var="linkTitleAttr">${%Took} ${build.durationString}</j:set>
           </j:if>
           <div tooltip="${linkTitleAttr}" data-tooltip-append-to-parent="true">
-            <i:formatDate value="${build.timestamp.time}" type="time" timeStyle="short" />
+            ${h.getDisambiguatedTime(build)}
```
[the remark this answers is not in the document]
Comment (SheikhSuhail19) [reply core/src/main/resources/hudson/widgets/HistoryWidget/entry.jelly:55]: Integrating localized terms directly within the FormatDateTag could indeed be complex.
I will explore this strategy further to assess its feasibility and effectiveness.

For now we can proceed with the current implementation of `getDisambiguatedTime`, which sufficiently addresses the time format disambiguation.

[the remark this answers is not in the document]
Comment (SheikhSuhail19) [reply core/src/main/resources/hudson/widgets/HistoryWidget/entry.jelly:55]: Latest build:
image

File: core/src/main/java/hudson/Functions.java (hunk 1 above)
Comment (SheikhSuhail19) [reply core/src/main/java/hudson/Functions.java]: FYI, while testing the suggested changes, I found that the time between 12:00:01 and 12:00:59 was not being handled correctly. I’ve addressed this by implementing truncation to the nearest minute, which ensures that any time between 12:00:00 and 12:00:59 (including 12:00:01) is correctly identified as Midnight or Noon.

This should now provide consistent and accurate labeling for times around midnight and noon.

Comment (SheikhSuhail19) [reply core/src/main/java/hudson/Functions.java]: Latest build:

**Review by A1exKH — APPROVED** [review]
LGTM.

File: core/src/main/resources/hudson/Messages_de.properties [hunk 3]
```diff
@@ -75,3 +75,6 @@ AboutJenkins.DisplayName=Über Jenkins
 AboutJenkins.Description=Versions- und Lizenzinformationen anzeigen.
 
 Functions.NoExceptionDetails=Keine Details zum Ausnahmefehler
+
+DisambiguatedTime.Midnight=0:00 Uhr
+DisambiguatedTime.Noon=12:00 Uhr
```
Comment (mawinter69) [inline core/src/main/resources/hudson/Messages_de.properties:80]: Here you're using the 24h format. Correct would be
```
12:00 Mitternacht
12:00 Mittag
```
But most likely this is unreachable anyway

**Review by mawinter69 — COMMENTED** [review]
Does it make sense to provide the 12h strings for noon and midnight for locales that use 24h format?
Seems you used google to translate the times as I doubt you're a native speaker of so many languages.

**Comment by jglick** [comment]
IMO the current behavior is not worth adjusting. If you are accustomed to 12h time formats, you should be accustomed to knowing what 12:00 AM and 12:00 PM mean in standard format strings. If you find these meanings confusing, switch to a 24h format. Anyway this would only apply to a build which started at that exact minute, which is not particularly likely; even if you are using a crontab, use the `H` syntax and most builds will not start exactly on the hour.

**Comment by krisstern** [comment]
Agreeing with what @jglick has just stated

**Comment by NotMyFault** [comment]
Given the former rationale given by Jesse and no further activity by the creator, I'll go ahead and close the PR.
